### PR TITLE
doc: add more details about memory domains

### PR DESCRIPTION
**Pull request by tyrantlizard77**
We now more throroughly discuss memory domains, thread
resource pools, and automatic memory domains.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

File: doc/reference/usermode/memory_domain.rst [hunk 1]
```diff
@@ -1,38 +1,288 @@
...
+- A read-only, executable region or regions for program text and ro-data, that
+  is accessible to user mode. This could be further sub-divided into a
+  read-only region for rodata, and a read-only, executable region for text, but
+  this will require an additional MPU region. This is required so that
+  threads running in user mode can read rodata and fetch instructions.
+
+Assuming a background map which allows supervisor mode to access any memory
+it needs, and regions which grant user mode access to text/rodata, this is
+sufficient for the boot time configuration.
+
```
Comment (ioannisg) [inline doc/reference/usermode/memory_domain.rst]: Certain configurations have also PRV-RW/nPRIV-RW fixed (static, i.e. programmed at boot) regions for GCOV, and HEP as well - do we need to mention here?

File: doc/reference/usermode/memory_domain.rst [hunk 2]
```diff
@@ -1,38 +1,288 @@
...
+Hardware Stack Overflow
+***********************
+
+CONFIG_HW_STACK_PROTECTION is an optional feature which detects stack
+buffer overflows when the system is running in supervisor mode. This
+catches issues when the entire stack buffer has overflowed, and not
+individual stack frames, use compiler-assisted CONFIG_STACK_CANARIES for that.
+
+Like any crash in supervisor mode, no guarantees can be made about the overall
```
Comment (ioannisg) [inline doc/reference/usermode/memory_domain.rst]: Should we stress that the feature guarantees, at least, the detection?

File: doc/reference/usermode/memory_domain.rst [hunk 3]
```diff
@@ -1,38 +1,288 @@
...
+and no extra MPU regions will be required.
+
+Thread Stack
+************
+
+Any thread running in user mode will need access to its own stack buffer.
+On context switch into a user mode thread, a dedicated MPU region will be
+programmed with the bounds of the stack buffer. A thread exceeding its stack
+buffer will start pushing data onto memory it doesn't have access to and an
+exception will be generated.
```
Comment (ioannisg) [inline doc/reference/usermode/memory_domain.rst]: memory access violation exception

File: doc/reference/usermode/memory_domain.rst [hunk 4]
```diff
@@ -1,38 +1,288 @@
...
+
+Thread Resource Pools
+*********************
+
+A small subset of kernel APIs require heap memory allocations which invoked
+as system calls. This memory is used only by the kernel and is not accessible
+directly to user mode. In order to use these system calls, invoking threads
```
Comment (ioannisg) [inline doc/reference/usermode/memory_domain.rst]: by user mode

**Review by ioannisg — APPROVED** [review]
I did a pass, this really looks nice!
I left some comments / suggestions to improve, please, take a look

**Comment by ioannisg** [comment]
@andrewboie does this PR address entirely #14206

**Comment by andrewboie** [comment]
> @andrewboie does this PR address entirely #14206??

It doesn't. It does tick some of the boxes. I'm going to keep working on it in 2.1.

File: doc/reference/usermode/memory_domain.rst [hunk 5]
```diff
@@ -1,38 +1,294 @@
...
+read-write to supervisor mode. A great deal of care must be taken when working
+with such regions to not unintentionally cause the kernel to crash when
+accessing such a region.
+
+Memory domain APIs are only available to supervisor mode. The only control
+user mode has over memory domains is that any user thread's child threads
+will automatically become members of the parent's domain.
+
+Memory Partitions
+-----------------
```
Comment (dbkinder) [inline doc/reference/usermode/memory_domain.rst]: Our project [heading hierarchy convention](https://docs.zephyrproject.org/latest/guides/documentation/index.html#headings) is  ``# * = - ~``, so if this heading is a subheading of Memory Domains, it should use ``=====``s and if it's on the same level as Memory Domains, it should use ``******``s.  Go through the heading underlines from here down and verify they're at the level you're expecting.  (Check out the [build artifacts for this PR](https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/18321/docs/reference/usermode/memory_domain.html) to see what the generated doc looks like.)

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by andrewboie** [comment]
@dbkinder revised, can you take a look?

**Review by dbkinder — APPROVED** [review]
Looks great, thanks!

**Pull request merged**